rust/xaynet-sdk/src/state_machine/state_machine.rs

```diff
@@ -0,0 +1,118 @@
+use derive_more::From;
+
+use crate::{settings::PetSettings, ModelStore, Notify, XaynetClient};
+
+use super::{
+    boxed_io,
+    Awaiting,
+    NewRound,
+    Phase,
+    SerializableState,
+    SharedState,
+    State,
+    Sum,
+    Sum2,
+    Update,
+};
+
+/// Outcome of a state machine transition attempt.
+pub enum TransitionOutcome {
+    /// Outcome when the state machine cannot make immediate
+    /// progress. The state machine is returned state unchanged.
+    Pending(StateMachine),
+    /// Outcome when a transition occured and the state machine was
+    /// updated.
+    Complete(StateMachine),
+}
+
+/// PET state machine.
+#[derive(From)]
+pub enum StateMachine {
+    /// PET state machine in the "new round" phase
+    NewRound(Phase<NewRound>),
+    /// PET state machine in the "awaiting" phase
+    Awaiting(Phase<Awaiting>),
+    /// PET state machine in the "sum" phase
+    Sum(Phase<Sum>),
+    /// PET state machine in the "update" phase
+    Update(Phase<Update>),
+    /// PET state machine in the "sum2" phase
+    Sum2(Phase<Sum2>),
+}
+
+impl StateMachine {
+    /// Try to make progress in the PET protocol
+    pub async fn transition(self) -> TransitionOutcome {
+        match self {
+            StateMachine::NewRound(phase) => phase.step().await,
+            StateMachine::Awaiting(phase) => phase.step().await,
+            StateMachine::Sum(phase) => phase.step().await,
+            StateMachine::Update(phase) => phase.step().await,
+            StateMachine::Sum2(phase) => phase.step().await,
+        }
+    }
+
+    /// Convert the state machine into a serializable datastructure so
```

Comment:
```suggestion
    /// Convert the state machine into a serializable data structure so
```

Reply:
Oh weird. I've always written that wrong.